helix-core/src/main/java/org/apache/helix/manager/zk/ZKUtil.java

```diff
@@ -92,15 +97,17 @@ public static boolean isClusterSetup(String clusterName, HelixZkClient zkClient)
     requiredPaths.add(PropertyPathBuilder.controllerHistory(clusterName));
     boolean isValid = true;
 
-    BaseDataAccessor<Object> baseAccessor = new ZkBaseDataAccessor<Object>(zkClient);
-    boolean[] ret = baseAccessor.exists(requiredPaths, 0);
+    boolean[] ret = new boolean[requiredPaths.size()];
+    for (int i = 0; i < requiredPaths.size(); i++) {
+      ret[i] = zkClient.exists(requiredPaths.get(i));
+    }
     StringBuilder errorMsg = new StringBuilder();
 
     for (int i = 0; i < ret.length; i++) {
       if (!ret[i]) {
         isValid = false;
-        errorMsg
-            .append(("Invalid cluster setup, missing znode path: " + requiredPaths.get(i)) + "\n");
+        errorMsg.append("Invalid cluster setup, missing znode path: ").append(requiredPaths.get(i))
```

Comment:
Nit: Maybe we can modify a bit for the error message to combine the error paths together like below to avoid the extra duplicate strings `Invalid cluster setup, missing znode path` for each path.
```
Invalid cluster setup, missing znode paths: [path1, path2]
```

Reply:
Feel free to do this in a future PR. This PR is not about improving or modifying the logic of ZkUtil. It simply is making things realm-aware.